query-parser/src/parse.rs

```diff
@@ -136,21 +178,25 @@ def_value_parser_fn!(Where, pattern, Pattern, input, {
             // ```
             //
             // is nonsense. That leaves us with nested optionals; we unwrap them here.
-            let r: ParseResult<Option<Pattern>, _> = p.parse_lazy(&y[..]).into();
-            let v: Option<Option<Pattern>> = r.ok().map(|x| x.0);
+            let r: ParseResult<Option<WhereClause>, _> = p.parse_lazy(&y[..]).into();
+            let v: Option<Option<WhereClause>> = r.ok().map(|x| x.0);
             v.unwrap_or(None)
         } else {
             None
         }
     }).parse_stream(input)
 });
 
+def_value_parser_fn!(Query, arguments, Vec<FnArg>, input, {
+    (many::<Vec<FnArg>, _>(Query::fn_arg()), eof())
+        .map(|(args, _)| { args })
+    .parse_stream(input)
+});
+
 def_value_parser_fn!(Where, clauses, Vec<WhereClause>, input, {
-    // Right now we only support patterns. See #239 for more.
-    (many1::<Vec<Pattern>, _>(Where::pattern()), eof())
-        .map(|(patterns, _)| {
-            patterns.into_iter().map(WhereClause::Pattern).collect()
-        })
+    // Right now we only support patterns and predicates. See #239 for more.
+    (many1::<Vec<WhereClause>, _>(choice([Where::pattern(), Where::pred()])), eof())
```

Comment:
There's a weirdness using `combine` where `choice` and parsers that consume some input  require care.  Have you taken that care here?  (I don't know the details offhand so I can't determine if you have without doing some research myself, so I'd like to save that research if you have done it.)